qiskit_experiments/tomography/tomography_experiment.py

```diff
@@ -0,0 +1,307 @@
+        num_meas = len(self._meas_qubits) if self._meas_qubits else self.num_qubits
+        meas_elements = product(range(meas_size), repeat=num_meas)
+        if self._prep_circ_basis:
+            prep_size = len(self._prep_circ_basis)
+            num_prep = len(self._prep_qubits) if self._prep_qubits else self.num_qubits
+            prep_elements = product(range(prep_size), repeat=num_prep)
+        else:
+            prep_elements = [None]
+
+        return product(meas_elements, prep_elements)
+
+    @staticmethod
+    def _permute_circuit(
+        circuit: QuantumCircuit,
+        measurement_qubits: Optional[Iterable[int]] = None,
+        preparation_qubits: Optional[Iterable[int]] = None,
+    ):
+        """Permute circuit qubits.
+
+        This permutes the circuit so that the specified preparation and measurement
+        qubits correspond to input and output qubits [0, ..., N-1] respectively
+        for the returned circuit.
+        """
+        if measurement_qubits is None and preparation_qubits is None:
+            return circuit
+
+        total_qubits = circuit.num_qubits
+        total_clbits = circuit.num_clbits
+        if total_clbits:
+            perm_circ = QuantumCircuit(total_qubits, total_clbits)
+        else:
+            perm_circ = QuantumCircuit(total_qubits)
+
+        # Apply permutation to put prep qubits as [0, ..., M-1]
+        if preparation_qubits:
+            prep_qargs = list(preparation_qubits)
+            if len(preparation_qubits) != total_qubits:
+                prep_qargs += [i for i in range(total_qubits) if i not in preparation_qubits]
+            perm_circ.append(Permutation(total_qubits, prep_qargs).inverse(), range(total_qubits))
+
+        # Apply original circuit
+        if total_clbits:
+            perm_circ = perm_circ.compose(circuit, range(total_qubits), range(total_clbits))
+        else:
+            perm_circ = perm_circ.compose(circuit, range(total_qubits))
+
+        # Apply permutation to put meas qubits as [0, ..., M-1]
+        if measurement_qubits:
+            meas_qargs = list(measurement_qubits)
+            if len(measurement_qubits) != total_qubits:
+                meas_qargs += [i for i in range(total_qubits) if i not in measurement_qubits]
+            perm_circ.append(Permutation(total_qubits, meas_qargs), range(total_qubits))
+
+        return perm_circ
+
+    @classmethod
+    def _target_quantum_state(
+        cls, circuit: QuantumCircuit, measurement_qubits: Optional[Iterable[int]] = None
+    ):
+        """Return the state tomography target"""
+        # Check if circuit contains measure instructions
+        # If so we cannot return target state
+        circuit_ops = circuit.count_ops()
+        if "measure" in circuit_ops:
+            return None
+
+        perm_circ = cls._permute_circuit(circuit, measurement_qubits=measurement_qubits)
+        if "reset" in circuit_ops or "kraus" in circuit_ops or "superop" in circuit_ops:
```

Comment:
Are they for simulators? I don't think these are executable with real backends.

Reply:
Since this function is only used to compute target state via simulation, this is checking ops in circuit so it knows what simulation method to use (DensityMatrix/Choi or Statevector/Operator).